### ISSUE TITLE
With XQuartz 2.8.5, drawing on the display is incomplete. 

### ISSUE THREAD

**Issue by pokynmr** [XQuartz/XQuartz#323]
I have tested a few Macs including XQuartz 2.8.5 and previous releases. I think the backing restore doesn't finish correctly on v. 2.8.5 because when I try to capture the desktop, it refreshes undrawn part and makes drawn. So I took a photo. 

![image](https://user-images.githubusercontent.com/46821344/216497092-a905c274-3494-4c85-9127-2034ddc434b0.png)

2.8.4 works fine.

Thanks,
POKY

**Comment by jeremyhu** [comment]
Can you please identify which prerelease of 2.8.5 introduced your issue and provide steps for me to reproduce it?

**Comment by pokynmr** [comment]
It is the official 2.8.5 bumped up from 2.8.4. The program is found from here: https://poky.clas.ucdenver.edu/poky_releases/current/poky_mac.zip

To reproduce, you can go menu Ext. -> Spectrum -> Simulate 2D Spectrum with BMRB -> Click N-HSQC and click OK with 6457 and specify where to save the simulated spectrum file when the BMRB entry number is asked. 

You can zoom in and out by + and - keys (or typing "zi" and "zo") and navigate using arrow keys. You will see that labels and some parts of contour lines randomly appear and disappear. It didn't happen from 2.8.4.

Thank you.
POKY

**Comment by jeremyhu** [comment]
Yes, I understand that you are seeing this from 2.8.4 to 2.8.5.  A lot changed between 2.8.4 and 2.8.5, and I was very deliberate in how I staged those changes across the betas in order to better identify what might have caused regressions.  Can you please idenfity which pre-release introduced the regression?  You can see all the pre-releases at https://www.xquartz.org/releases/archive.html or in the releases section in github.

**Comment by pokynmr** [comment]
I've been through all versions. I see this problem with 2.8.5_beta3 and newer.

**Comment by jmw86069** [comment]
If it helps, I confirm the regression occurs on 2.8.5_beta3, but not 2.8.5_beta2 or prior version 2.8.4.

My symptom is that display is only partially rendered. After moving the window (clicking header, dragging somewhere) more features are rendered, but is still not completely rendered. With version 2.8.4, and 2.8.5_beta2 these behaviors are not observed.

Good luck with tracking down the issue! The description of beta3 seemed relevant to what we are experiencing, so hopefully it's something you recognize when you go through it. And thanks for your time!

**Comment by biobee** [comment]
Dear all,

I experience a drawing issue with NEdit on XQuartz 2.8.5. (Intel Mac, Monterey 12.6.3).  I am now back at 2.8.4 which works fine. The remote NEdit is released by Debian (1:5.7-2).

The issue: the Menu items of NEdit are not drawn when the window appears, but only after resizing the NEdit window.
NEdit menus are:
File Edit Search Preferences etc, with _ under the text for shortcuts. When the window is initially build on screen only the _ appear, but not the text above it. 

The normal NEdit menu is attached. With the new XQuartz version, only all  _ appear, until a window resize.


![NEdit_menu](https://user-images.githubusercontent.com/48982323/217617510-44268501-5555-48e2-ba1d-4fed81f28575.jpg)

**Comment by jmw86069** [comment]
Honestly, this release is a major regression from 2.8.4 and it really should be removed. It is not functional, and I don't understand having a full release that is not functional at all. Maybe there is some other benefit to version 2.8.5 over 2.8.4 that suggests keeping it online. Otherwise it seems like a regression, it happens, pull it and wait for a fix.

**Comment by jmw86069** [comment]
I just wanted to thank you for the response, given the state of things that was a skillfully worded and respectful response. I wouldn't blame you for being much more frustrated offline (haha), totally understandable. I was not aware this issue only affected a subset of users like me.

I'm not sure I have the expertise to dig into the codebase, but ya never know. Would the best place to start be to look at the specific pre-release that caused the error? You mentioned that some old codebase was removed which presumably caused the decades-old bugs to re-emerge. I don't know if it more helpful to narrow down the very specific use case(s) that cause the problem, or to review the codebase itself.

**Comment by jeremyhu** [comment]
Yeah, if you look at the previous release's commits (https://github.com/XQuartz/xorg-server/commits/XQuartz-2.8.4), we had a bunch of "use old ..." "revert ..." commits that we were rebasing for a while.

In 2.8.5, those reverts were removed, and I made a bunch of changes to that subsystem (rootless) to address as much of the core problem as I could.

There are also links to the old bug reports on the issue which might have some info:

https://github.com/XQuartz/xquartz-old-tickets/blob/master/ticket/525.md
http://bugs.freedesktop.org/26124

**Comment by gordonj7** [comment]
I posted a similar ticket #351 that was closed as a dupe of this.

Just wanted to add my two cents that this issue does not occur when accessing X apps on an older RHEL6 host.  We only noticed it when migrating to RHEL9.

So perhaps the underlying bug is some combination of "new XQuartz" and "new X server".

**Comment by jeremyhu** [comment]
@pokynmr 

```
$ ./poky.app/Contents/MacOS/poky
dyld[4637]: Library not loaded: @loader_path/../Frameworks/libpython3.7m.dylib
  Referenced from: <CBBC47D3-8303-374D-9E0F-71E4526AE3EE> /Users/jeremy/Downloads/poky_mac/poky.app/Contents/MacOS/poky
  Reason: tried: '/Users/jeremy/Downloads/poky_mac/poky.app/Contents/MacOS/../Frameworks/libpython3.7m.dylib' (no such file), '/usr/local/lib/libpython3.7m.dylib' (no such file), '/usr/lib/libpython3.7m.dylib' (no such file, not in dyld cache)
zsh: abort      ./poky.app/Contents/MacOS/poky
```

Looks lke your app bundle is missing libpython3.7m.dylib, so I'm not able to check if it's actually fixed with these changes.

**Comment by pokynmr** [comment]
> [@pokynmr](https://github.com/pokynmr)
> 
> ```
> $ ./poky.app/Contents/MacOS/poky
> dyld[4637]: Library not loaded: @loader_path/../Frameworks/libpython3.7m.dylib
>   Referenced from: <CBBC47D3-8303-374D-9E0F-71E4526AE3EE> /Users/jeremy/Downloads/poky_mac/poky.app/Contents/MacOS/poky
>   Reason: tried: '/Users/jeremy/Downloads/poky_mac/poky.app/Contents/MacOS/../Frameworks/libpython3.7m.dylib' (no such file), '/usr/local/lib/libpython3.7m.dylib' (no such file), '/usr/lib/libpython3.7m.dylib' (no such file, not in dyld cache)
> zsh: abort      ./poky.app/Contents/MacOS/poky
> ```
> 
> Looks lke your app bundle is missing libpython3.7m.dylib, so I'm not able to check if it's actually fixed with these changes.

<img width="997" height="327" alt="Image" src="https://github.com/user-attachments/assets/c4f56b28-c077-4346-aac3-0ad6be076e11" />

I just downloaded and checked, and it is there.

**Comment by qwertychouskie** [comment]
> > [@pokynmr](https://github.com/pokynmr)
> > ```
> > $ ./poky.app/Contents/MacOS/poky
> > dyld[4637]: Library not loaded: @loader_path/../Frameworks/libpython3.7m.dylib
> >   Referenced from: <CBBC47D3-8303-374D-9E0F-71E4526AE3EE> /Users/jeremy/Downloads/poky_mac/poky.app/Contents/MacOS/poky
> >   Reason: tried: '/Users/jeremy/Downloads/poky_mac/poky.app/Contents/MacOS/../Frameworks/libpython3.7m.dylib' (no such file), '/usr/local/lib/libpython3.7m.dylib' (no such file), '/usr/lib/libpython3.7m.dylib' (no such file, not in dyld cache)
> > zsh: abort      ./poky.app/Contents/MacOS/poky
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Looks lke your app bundle is missing libpython3.7m.dylib, so I'm not able to check if it's actually fixed with these changes.
> 
> <img alt="Image" width="997" height="327" src="https://private-user-images.githubusercontent.com/46821344/567465795-c4f56b28-c077-4346-aac3-0ad6be076e11.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzQyNTkzNzQsIm5iZiI6MTc3NDI1OTA3NCwicGF0aCI6Ii80NjgyMTM0NC81Njc0NjU3OTUtYzRmNTZiMjgtYzA3Ny00MzQ2LWFhYzMtMGFkNmJlMDc2ZTExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAzMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMzIzVDA5NDQzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThhN2UxMzhkYTA2OTc5NjdkM2Y1YmJhNDA1ZWUxNTNkNmM1NjljZmUyZGE2NjYxN2JlOGU5OWM2OTg3MjUyZWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5FJA5fq0o5z2SIIYkul8HwJsySVN5lzcUyjwTLDD8mY">
> 
> I just downloaded and checked, and it is there.

I know arm64 macOS is more strict about code signing and stuff for libraries, so `libpython3.7m.dylib` might be present, but not allowed to load, due to e.g. a mismatched signature to the rest of the bundle or something like that.  ([I ran into an issue like this with SuperTux](https://github.com/SuperTux/supertux/pull/3499).)

**Comment by jeremyhu** [comment]
```
~/Downloads/poky_mac $ ls poky.app/Contents/Frameworks
Qt3DAnimation.framework       QtBluetooth.framework          QtMultimediaQuick.framework    QtQmlWorkerScript.framework       QtQuickWidgets.framework     QtWebEngine.framework
Qt3DCore.framework            QtBodymovin.framework          QtMultimediaWidgets.framework  QtQuick.framework                 QtRemoteObjects.framework    QtWebEngineCore.framework
Qt3DExtras.framework          QtCharts.framework             QtNetwork.framework            QtQuick3D.framework               QtScxml.framework            QtWebEngineWidgets.framework
Qt3DInput.framework           QtConcurrent.framework         QtNfc.framework                QtQuick3DAssetImport.framework    QtSensors.framework          QtWebSockets.framework
Qt3DLogic.framework           QtCore.framework               QtOpenGL.framework             QtQuick3DRender.framework         QtSerialBus.framework        QtWebView.framework
Qt3DQuick.framework           QtDataVisualization.framework  QtPdf.framework                QtQuick3DRuntimeRender.framework  QtSerialPort.framework       QtWidgets.framework
Qt3DQuickAnimation.framework  QtDBus.framework               QtPositioning.framework        QtQuick3DUtils.framework          QtSql.framework              QtXml.framework
Qt3DQuickExtras.framework     QtDesigner.framework           QtPositioningQuick.framework   QtQuickControls2.framework        QtSvg.framework              QtXmlPatterns.framework
Qt3DQuickInput.framework      QtGamepad.framework            QtPrintSupport.framework       QtQuickParticles.framework        QtTest.framework             tcl-tk-x11
Qt3DQuickRender.framework     QtGui.framework                QtPurchasing.framework         QtQuickShapes.framework           QtTextToSpeech.framework
Qt3DQuickScene2D.framework    QtLocation.framework           QtQml.framework                QtQuickTemplates2.framework       QtVirtualKeyboard.framework
Qt3DRender.framework          QtMultimedia.framework         QtQmlModels.framework          QtQuickTest.framework             QtWebChannel.framework
```

=/

**Comment by pokynmr** [comment]
> ```
> ~/Downloads/poky_mac $ ls poky.app/Contents/Frameworks
> Qt3DAnimation.framework       QtBluetooth.framework          QtMultimediaQuick.framework    QtQmlWorkerScript.framework       QtQuickWidgets.framework     QtWebEngine.framework
> Qt3DCore.framework            QtBodymovin.framework          QtMultimediaWidgets.framework  QtQuick.framework                 QtRemoteObjects.framework    QtWebEngineCore.framework
> Qt3DExtras.framework          QtCharts.framework             QtNetwork.framework            QtQuick3D.framework               QtScxml.framework            QtWebEngineWidgets.framework
> Qt3DInput.framework           QtConcurrent.framework         QtNfc.framework                QtQuick3DAssetImport.framework    QtSensors.framework          QtWebSockets.framework
> Qt3DLogic.framework           QtCore.framework               QtOpenGL.framework             QtQuick3DRender.framework         QtSerialBus.framework        QtWebView.framework
> Qt3DQuick.framework           QtDataVisualization.framework  QtPdf.framework                QtQuick3DRuntimeRender.framework  QtSerialPort.framework       QtWidgets.framework
> Qt3DQuickAnimation.framework  QtDBus.framework               QtPositioning.framework        QtQuick3DUtils.framework          QtSql.framework              QtXml.framework
> Qt3DQuickExtras.framework     QtDesigner.framework           QtPositioningQuick.framework   QtQuickControls2.framework        QtSvg.framework              QtXmlPatterns.framework
> Qt3DQuickInput.framework      QtGamepad.framework            QtPrintSupport.framework       QtQuickParticles.framework        QtTest.framework             tcl-tk-x11
> Qt3DQuickRender.framework     QtGui.framework                QtPurchasing.framework         QtQuickShapes.framework           QtTextToSpeech.framework
> Qt3DQuickScene2D.framework    QtLocation.framework           QtQml.framework                QtQuickTemplates2.framework       QtVirtualKeyboard.framework
> Qt3DRender.framework          QtMultimedia.framework         QtQmlModels.framework          QtQuickTest.framework             QtWebChannel.framework
> ```
> 
> =/

Well, that's weird. 
$ unzip -l poky_mac.zip | grep python3.7m
    49704  11-20-2025 16:19   poky.app/Contents/Resources/bin/python3.7m
 13566464  12-23-2021 22:04   poky.app/Contents/Resources/python3.7/config-3.7m-darwin/libpython3.7m.a
  2552584  11-20-2025 16:19   poky.app/Contents/Frameworks/libpython3.7m.dylib

But, nvm. I confirm this issue is addressed in 2.8.6.b3. Thanks!
